Expand handler sets and add handler stacks (#2877)
sensu/sensu-docs: content/sensu-go/6.1/observability-pipeline/observe-process/handlers.md

## Handler sets

Handler set definitions allow you to use a single named handler set to refer to groups of handlers.
The handler set becomes a collection of individual actions to take (via each included handler) on event data.

For example, suppose you have already created these two handlers:

- `elasticsearch` to send all observation data to Elasticsearch.
- `opsgenie` to send non-OK status alerts to your OpsGenie notification channel.

You can list both of these handlers in a handler set to automate and streamline your workflow, specifying `type: set`:

{{< language-toggle >}}

{{< code yml >}}
type: Handler
api_version: core/v2
metadata:
  name: send_events_notify_operator
  namespace: default
spec:
  handlers:
  - elasticsearch
  - opsgenie
  type: set
{{< /code >}}

{{< code json >}}
{
  "type": "Handler",
  "api_version": "core/v2",
  "metadata": {
    "name": "send_events_notify_operator",
    "namespace": "default"
  },
  "spec": {
    "type": "set",
    "handlers": [
      "elasticsearch",
      "opsgenie"
    ]
  }
}
{{< /code >}}

{{< /language-toggle >}}

> Now you can route observation data to Elasticsearch and alerts to OpsGenie with a single handler, the `send_events_notify_operator` handler set.

Reviewer: Should this maybe say "[...] with a single handler definition, the `send_events_notify_operator` handler set."?

Author: ```suggestion
Now you can route observation data to Elasticsearch and alerts to OpsGenie with a single handler definition, the `send_events_notify_operator` handler set.
```

{{% notice note %}}
**NOTE**: Attributes defined in handler sets do not apply to the handlers they include.
For example, `filters` and `mutator` attributes defined in a handler set will have no effect on handlers.
Define these attributes in individual handlers instead.
{{% /notice %}}

## Handler stacks

> Handler stacks are handler sets that escalate events through a series of different handlers.

Reviewer: The stacking can be done independent of handler sets.  Maybe re-write this as "Hander stacks are a group of handlers, or a handler set, that escalate events [...]"?

> Handler stacks are handler sets that escalate events through a series of different handlers.

Author: ```suggestion
Handler stacks are a group of handlers or a handler set that escalate events through a series of different handlers.
```

For example, consider a handler stack with three levels of escalation:

- Level 1: On first occurrence, attempt remediation with the `remediation` handler.
- Level 2: On fifth occurrence, send an alert to Slack with the `slack_alert_channel` handler.
- Level 3: On tenth occurrence, send an alert to PagerDuty with the `pagerduty_emergency` handler.

The escalation levels are controlled by event filter definitions.

> At level 1, no filter is needed because every event triggers remediation with the `remediation` handler.

Reviewer: No filter needed is not entirely correct.  You would still want the `is_incident` filter otherwise on the initial successful check the remediation would still be ran.

> At level 1, no filter is needed because every event triggers remediation with the `remediation` handler.

Author: ```suggestion
At level 1, you only need the [`is_incident` event filter][31].
Every event triggers remediation with the `remediation` handler.
```

      "opsgenie"
    ]
  }
}
{{< /code >}}

{{< /language-toggle >}}

Now you can route observation data to Elasticsearch and alerts to OpsGenie with a single handler definition, the `send_events_notify_operator` handler set.

{{% notice note %}}
**NOTE**: Attributes defined in handler sets do not apply to the handlers they include.
For example, `filters` and `mutator` attributes defined in a handler set will have no effect on handlers.
Define these attributes in individual handlers instead.
{{% /notice %}}

## Handler stacks

Handler stacks are a group of handlers or a handler set that escalate events through a series of different handlers.
For example, consider a handler stack with three levels of escalation:

- Level 1: On first occurrence, attempt remediation with the `remediation` handler.
- Level 2: On fifth occurrence, send an alert to Slack with the `slack_alert_channel` handler.
- Level 3: On tenth occurrence, send an alert to PagerDuty with the `pagerduty_emergency` handler.

The escalation levels are controlled by event filter definitions.
At level 1, you only need the [`is_incident` event filter][31].
Every event triggers remediation with the `remediation` handler.

> For escalation to levels 2 and 3, you can define an event filter that uses the [fatigue check filter][30] asset to activate the `slack_alert_channel` and `pagerduty_emergency` handlers at the fifth and tenth occurrences, respectively. 

Reviewer: In thinking about this further.  You would want `is_incident` filtering on all handlers, in addition to occurrence filter either directly (link to something regarding occurrences filtering) or via fatigue check filter.

I also wonder if we should make note that certain remediation handlers (the Sensu provided Ansible, Rundeck, Saltstack, and generic remediation handlers) have their own built-in occurrences and severities filtering.

      "elasticsearch",
      "opsgenie"
    ]
  }
}
{{< /code >}}

{{< /language-toggle >}}

Now you can route observation data to Elasticsearch and alerts to OpsGenie with a single handler definition, the `send_events_notify_operator` handler set.

{{% notice note %}}
**NOTE**: Attributes defined in handler sets do not apply to the handlers they include.
For example, `filters` and `mutator` attributes defined in a handler set will have no effect on handlers.
Define these attributes in individual handlers instead.
{{% /notice %}}

## Handler stacks

Handler stacks are a group of handlers or a handler set that escalate events through a series of different handlers.
For example, suppose you want a handler stack with three levels of escalation:

- Level 1: On the first occurrence, attempt remediation.
- Level 2: On the fifth occurrence, send an alert to Slack.
- Level 3: On the tenth occurrence, send an alert to PagerDuty.
Continue to send this alert on every tenth occurrence thereafter until the incident is resolved.

A handler stack for this scenario requires three handlers to take the desired actions based on three corresponding event filters that control the escalation levels:

> - Level 1 requires an event filter with the built-in [`is_incident` filter][30] and a corresponding remediation handler.

Reviewer: This will also require an `occurrences == 1 filter` if **NOT** using a remediation handler that does its own occurrences checking.

> - Level 1 requires an event filter with the built-in [`is_incident` filter][30] and a corresponding remediation handler.

Author: ```suggestion
- Level 1 requires an event filter with the built-in [`is_incident` filter][30] plus an [occurrence-based filter][32] that uses an expression like `event.check.occurrences ==1` and a corresponding remediation handler.
```

> - Level 2 requires an event filter with `is_incident` plus an [occurrence-based filter][32] that uses an expression like `event.check.occurrences == 5` and a corresponding Slack handler.
> - Level 3 requires an event filter with `is_incident` plus an [occurrence-based filter][32] that uses an expression like `event.check.occurrences % 10 == 0` to match event data with an occurrences value that is evenly divisible by 10 via a modulo operator calculation and a corresponding PagerDuty handler.

Author: ```suggestion
- Level 2 requires an event filter with `is_incident` plus an occurrence-based filter that uses an expression like `event.check.occurrences == 5` and a corresponding Slack handler.
- Level 3 requires an event filter with `is_incident` plus an occurrence-based filter that uses an expression like `event.check.occurrences % 10 == 0` to match event data with an occurrences value that is evenly divisible by 10 via a modulo operator calculation and a corresponding PagerDuty handler.
```

[27]: ../../observe-schedule/agent/#registration-endpoint-management-and-service-discovery
[28]: ../../../web-ui/search/
[29]: ../../../observability-pipeline/

> [30]: https://bonsai.sensu.io/assets/nixwiz/sensu-go-fatigue-check-filter

Author: ```suggestion
[30]: https://bonsai.sensu.io/assets/nixwiz/sensu-go-fatigue-check-filter
[31]: ../../observe-filter/filters/#built-in-filter-is_incident

```
